lxd/storage/drivers/driver_alletra_volumes.go

```diff
@@ -75,6 +85,286 @@ func (d *alletra) getVolumeName(vol Volume) (string, error) {
+// mode included.
+func (d *alletra) ensureHost() (hostName string, cleanup revert.Hook, err error) {
+	var hostname string
+
+	revert := revert.New()
+	defer revert.Fail()
+
+	connector, err := d.connector()
+	if err != nil {
+		return "", nil, err
+	}
+
+	// Get the qualified name of the host.
+	qn, err := connector.QualifiedName()
+	if err != nil {
+		return "", nil, err
+	}
+
+	// Fetch an existing host entry on a storage array.
+	host, err := d.client().GetCurrentHost(connector.Type(), qn)
+	if err != nil {
+		if !api.StatusErrorCheck(err, http.StatusNotFound) {
+			return "", nil, err
+		}
+
+		// The storage host entry with a qualified name of the current LXD host does not exist.
+		// Therefore, create a new one and name it after the server name.
+		serverName, err := ResolveServerName(d.state.ServerName)
+		if err != nil {
+			return "", nil, err
+		}
+
+		// Append the mode to the server name because storage array does not allow mixing
+		// NQNs, IQNs, and WWNs for a single host.
+		hostname = serverName + "-" + connector.Type()
+
+		err = d.client().CreateHost(connector.Type(), hostname, []string{qn})
+		if err != nil {
+			if !api.StatusErrorCheck(err, http.StatusConflict) {
+				return "", nil, err
+			}
+
+			// The host with the given name already exists, update it instead.
+			err = d.client().UpdateHost(hostname, []string{qn})
+			if err != nil {
+				return "", nil, err
+			}
+		} else {
+			revert.Add(func() {
+				err := d.client().DeleteHost(hostname)
+				if err != nil {
+					d.logger.Error("DeleteHost API call failed on error path", logger.Ctx{"err": err, "hostname": hostname})
+				}
+			})
+		}
+	} else {
+		// Hostname already exists with the given IQN.
+		hostname = host.Name
+	}
+
+	cleanup = revert.Clone().Fail
+	revert.Success()
+	return hostname, cleanup, nil
+}
+
+// mapVolume maps the given volume onto this host.
+func (d *alletra) mapVolume(vol Volume) (cleanup revert.Hook, err error) {
+	reverter := revert.New()
+	defer reverter.Fail()
+
+	connector, err := d.connector()
+	if err != nil {
+		return nil, err
+	}
+
+	volName, err := d.getVolumeName(vol)
+	if err != nil {
+		return nil, err
+	}
+
+	unlock, err := remoteVolumeMapLock(connector.Type(), d.Info().Name)
+	if err != nil {
+		return nil, err
+	}
+
+	defer unlock()
+
+	// Ensure the host exists and is configured with the correct QN.
+	hostname, cleanup, err := d.ensureHost()
+	if err != nil {
+		return nil, err
+	}
+
+	reverter.Add(cleanup)
+
+	// Ensure the volume is connected to the host.
+	connCreated, err := d.client().ConnectHostToVolume(vol.pool, volName, hostname)
+	if err != nil {
+		return nil, err
+	}
+
+	if connCreated {
+		reverter.Add(func() {
+			err := d.client().DisconnectHostFromVolume(vol.pool, volName, hostname)
+			if err != nil {
+				d.logger.Error("DisconnectHostFromVolume API call failed on error path", logger.Ctx{"err": err, "volName": volName, "hostname": hostname})
```

Comment:
Suggest using `d.logger.Warn` here so as not to distract from the main error that is happening.

Reply:
Fixed, thanks!